service/history/nDCUpdateTransactionMgr.go

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	ctx "context"
+	"fmt"
+	"time"
+
+	"github.com/uber/cadence/.gen/go/shared"
+)
+
+type (
+	nDCUpdateTransactionMgr interface {
+		dispatchWorkflowUpdate(
+			ctx ctx.Context,
+			now time.Time,
+			isWorkflowRebuilt bool,
+			targetWorkflow nDCWorkflow,
+			newWorkflow nDCWorkflow,
+		) error
+	}
+
+	nDCUpdateTransactionMgrImpl struct {
+		transactionMgr *nDCTransactionMgrImpl
+	}
+)
+
+func newNDCUpdateTransactionMgr(
+	transactionMgr *nDCTransactionMgrImpl,
+) *nDCUpdateTransactionMgrImpl {
+
+	return &nDCUpdateTransactionMgrImpl{
+		transactionMgr: transactionMgr,
+	}
+}
+
+func (r *nDCUpdateTransactionMgrImpl) dispatchWorkflowUpdate(
+	ctx ctx.Context,
+	now time.Time,
+	isWorkflowRebuilt bool,
+	targetWorkflow nDCWorkflow,
+	newWorkflow nDCWorkflow,
+) error {
+
+	// NOTE: this function does NOT mutate current workflow, target workflow or new workflow,
+	//  workflow mutation is done in methods within executeTransaction function
+
+	// this is a performance optimization so most update does not need to
+	// check whether target workflow is current workflow by calling DB API
+	if targetWorkflow.getMutableState().IsCurrentWorkflowGuaranteed() {
+		// update to current record, since target workflow is pointed by current record
+		return r.dispatchWorkflowUpdateAsCurrent(
+			ctx,
+			now,
+			isWorkflowRebuilt,
+			targetWorkflow,
+			newWorkflow,
```

Comment:
Is it only for continue as new?

Reply:
nop, continue as new is handled as part of applying replication task, except when apply replication task to workflow's non current branch